README.md
[![Published on NPM](https://img.shields.io/npm/v/@advanced-rest-client/history-menu.svg)](https://www.npmjs.com/package/@advanced-rest-client/history-menu)

[![Build Status](https://travis-ci.org/advanced-rest-client/history-menu.svg?branch=stage)](https://travis-ci.org/advanced-rest-client/history-menu)

[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/advanced-rest-client/history-menu)

# &lt;history-menu&gt;

The history menu for Advanced REST Client is accessible list of history items retrieved from local data store.

## Deprecation notice

This project is no longer maintained. Use `@advanced-rest-client/arc-menu` instead.

## Usage

### Installation
```
npm install --save @advanced-rest-client/history-menu
```

### In a LitElement

```js
import { LitElement, html } from 'lit-element';
import '@advanced-rest-client/history-menu/history-menu.js';

class SampleElement extends LitElement {
  get styles() {
    return css`
      history-menu {
        height: 500px;
      }
    `;
  }

  render() {
    return html`
    <history-menu draggableenabled></history-menu>
    `;
  }
}
customElements.define('sample-element', SampleElement);
```

### List sizing

It is important to set implicit height of the element. It can be static value like `500px`, relative value like `100%` as long as a parent is sized for height, or a flex value, as long as parent is sized for height.
The list of requests is set to load only portion of the requests from the data store and load more when list scroll is near end. If there's no scroll then the element will load whole data store at initialization time.

### Drag and drop

API components related to a request object support drag and drop. Set `draggableenabled` property to enable the support.

The `DataTransfer` property of the drag event contains `effectAllowed` set to `copy` as this is only allowed operation on a history object. Only targets that allow the same effect will accept the history item.
The same propery contains serialized request data under `arc/request-object` media type. It contains request ID under `arc/history-request` and `arc-source/history-menu` media types.

```javascript
_dropHandler(e) {
  e.preventDefault();
  const data = e.dataTransfer.getData('arc/request-object');
  const request = JSON.parse(data);
  const id = e.dataTransfer.getData('arc/history-request');
  console.log(request, id);
}
```

## Development

```sh
git clone https://github.com/advanced-rest-client/history-list-mixin
cd history-list-mixin
npm install
```

### Running the tests

```sh
npm test
```

## API components

This components is a part of [API components ecosystem](https://elements.advancedrestclient.com/)


CHANGELOG.md
<a name="2.0.0"></a>
# [2.0.0](https://github.com/advanced-rest-client/history-menu/compare/0.1.7...2.0.0) (2019-01-10)


### Docs

* Updated styles docs ([11a858cd9cec03a07f8dd144bbce5094f5dcd058](https://github.com/advanced-rest-client/history-menu/commit/11a858cd9cec03a07f8dd144bbce5094f5dcd058))

### Update

* Adding requests lists mixin and styles ([f6a41207d7a0a25d40048f34db9f5b18151fbf45](https://github.com/advanced-rest-client/history-menu/commit/f6a41207d7a0a25d40048f34db9f5b18151fbf45))



<a name="0.1.8"></a>
## [0.1.8](https://github.com/advanced-rest-client/history-menu/compare/0.1.7...0.1.8) (2018-01-30)




<a name="0.1.7"></a>
## [0.1.7](https://github.com/advanced-rest-client/history-menu/compare/0.1.6...0.1.7) (2018-01-30)




<a name="0.1.6"></a>
## [0.1.6](https://github.com/advanced-rest-client/history-menu/compare/0.1.5...0.1.6) (2018-01-30)


### Update

* Adding CSS variables for empty info screen. ([3e8c1e67f66d35fc16533d6adc50c3e121bcd704](https://github.com/advanced-rest-client/history-menu/commit/3e8c1e67f66d35fc16533d6adc50c3e121bcd704))



<a name="0.1.5"></a>
## [0.1.5](https://github.com/advanced-rest-client/history-menu/compare/0.1.4...0.1.5) (2018-01-30)


### Update

* Adding CSS variables for empty info screen. ([66ebceb22de07cf83b08111e184b7d5956a079c9](https://github.com/advanced-rest-client/history-menu/commit/66ebceb22de07cf83b08111e184b7d5956a079c9))



<a name="0.1.4"></a>
## [0.1.4](https://github.com/advanced-rest-client/history-menu/compare/0.1.3...0.1.4) (2017-10-02)




<a name="0.1.3"></a>
## [0.1.3](https://github.com/advanced-rest-client/history-menu/compare/0.1.2...0.1.3) (2017-10-02)


### Update

* Adding hidden object css ([041c1c3c7b3001dd74a64ba8d22247c82737df90](https://github.com/advanced-rest-client/history-menu/commit/041c1c3c7b3001dd74a64ba8d22247c82737df90))



<a name="0.1.2"></a>
## 0.1.2 (2017-10-02)


### Update

* Added handler for all database destroy event. ([6e1ce172461190d2c745301e49ef6656dfd4c96e](https://github.com/advanced-rest-client/history-menu/commit/6e1ce172461190d2c745301e49ef6656dfd4c96e))
* Added sauce configuration for tests ([b7c3b9a5a7d3c18e798c59886d0ea0a7e8a03c83](https://github.com/advanced-rest-client/history-menu/commit/b7c3b9a5a7d3c18e798c59886d0ea0a7e8a03c83))
* Removed obsolite browsers from tests. ([f45d7be746fc33f2712ba6b7e8113f1403c8cdcb](https://github.com/advanced-rest-client/history-menu/commit/f45d7be746fc33f2712ba6b7e8113f1403c8cdcb))



<a name="0.1.1"></a>
## 0.1.1 (2017-09-13)


### Update

* Added sauce configuration for tests ([b7c3b9a5a7d3c18e798c59886d0ea0a7e8a03c83](https://github.com/advanced-rest-client/history-menu/commit/b7c3b9a5a7d3c18e798c59886d0ea0a7e8a03c83))
* Removed obsolite browsers from tests. ([f45d7be746fc33f2712ba6b7e8113f1403c8cdcb](https://github.com/advanced-rest-client/history-menu/commit/f45d7be746fc33f2712ba6b7e8113f1403c8cdcb))



## [2.0.1](https://github.com/advanced-rest-client/history-menu/compare/0.1.7...2.0.1) (2019-01-26)


### Docs

* Updated styles docs ([11a858cd9cec03a07f8dd144bbce5094f5dcd058](https://github.com/advanced-rest-client/history-menu/commit/11a858cd9cec03a07f8dd144bbce5094f5dcd058))

### Update

* Adding requests lists mixin and styles ([f6a41207d7a0a25d40048f34db9f5b18151fbf45](https://github.com/advanced-rest-client/history-menu/commit/f6a41207d7a0a25d40048f34db9f5b18151fbf45))
* Updating dependencies ([d50ad47f71ade2167fbf8c534dd08b1b1a15c43e](https://github.com/advanced-rest-client/history-menu/commit/d50ad47f71ade2167fbf8c534dd08b1b1a15c43e))



## [2.0.2](https://github.com/advanced-rest-client/history-menu/compare/0.1.7...2.0.2) (2019-02-09)


### Docs

* Updated styles docs ([11a858cd9cec03a07f8dd144bbce5094f5dcd058](https://github.com/advanced-rest-client/history-menu/commit/11a858cd9cec03a07f8dd144bbce5094f5dcd058))

### New

* Adding support for draggable action ([5ac17c6ba58f8277ee49e85f0cdcf929df6b5f5f](https://github.com/advanced-rest-client/history-menu/commit/5ac17c6ba58f8277ee49e85f0cdcf929df6b5f5f))

### Update

* Adding requests lists mixin and styles ([f6a41207d7a0a25d40048f34db9f5b18151fbf45](https://github.com/advanced-rest-client/history-menu/commit/f6a41207d7a0a25d40048f34db9f5b18151fbf45))
* Updating dependencies ([d50ad47f71ade2167fbf8c534dd08b1b1a15c43e](https://github.com/advanced-rest-client/history-menu/commit/d50ad47f71ade2167fbf8c534dd08b1b1a15c43e))
* Updating requests lsit version ([3b758f17686e38ed9964c3874a293d6344903b91](https://github.com/advanced-rest-client/history-menu/commit/3b758f17686e38ed9964c3874a293d6344903b91))



## [2.0.3](https://github.com/advanced-rest-client/history-menu/compare/0.1.7...2.0.3) (2019-02-09)


### Docs

* Updated styles docs ([11a858cd9cec03a07f8dd144bbce5094f5dcd058](https://github.com/advanced-rest-client/history-menu/commit/11a858cd9cec03a07f8dd144bbce5094f5dcd058))

### New

* Adding support for draggable action ([5ac17c6ba58f8277ee49e85f0cdcf929df6b5f5f](https://github.com/advanced-rest-client/history-menu/commit/5ac17c6ba58f8277ee49e85f0cdcf929df6b5f5f))

### Update

* Adding requests lists mixin and styles ([f6a41207d7a0a25d40048f34db9f5b18151fbf45](https://github.com/advanced-rest-client/history-menu/commit/f6a41207d7a0a25d40048f34db9f5b18151fbf45))
* Updating dependencies ([d50ad47f71ade2167fbf8c534dd08b1b1a15c43e](https://github.com/advanced-rest-client/history-menu/commit/d50ad47f71ade2167fbf8c534dd08b1b1a15c43e))
* Updating drag start to include more transfer data ([e87dd9b1ee81860dd30c554979eace701fc18e0a](https://github.com/advanced-rest-client/history-menu/commit/e87dd9b1ee81860dd30c554979eace701fc18e0a))
* Updating requests lsit version ([3b758f17686e38ed9964c3874a293d6344903b91](https://github.com/advanced-rest-client/history-menu/commit/3b758f17686e38ed9964c3874a293d6344903b91))



# [3.0.0](https://github.com/advanced-rest-client/history-menu/compare/0.1.7...3.0.0) (2019-09-06)




test/data-generator.js
/* global chance */
const DataGenerator = {};
const methods = ['GET', 'PUT', 'POST', 'DELETE', 'PATCH', 'HEAD'];
const methodsSize = methods.length - 1;

var LAST_TIME = Date.now();

function setMidninght(time) {
  var now = new Date(time);
  now.setMilliseconds(0);
  now.setSeconds(0);
  now.setMinutes(0);
  now.setHours(0);
  return now.getTime();
}

DataGenerator.genRequestObject = function() {
  var methodIndex = chance.integer({min: 0, max: methodsSize});
  var url = chance.url();
  LAST_TIME -= chance.integer({min: 1.8e+6, max: 8.64e+7});
  var time = LAST_TIME;
  var method = methods[methodIndex];
  var obj = {
    _id: setMidninght(time) + '/' + encodeURIComponent(url) + '/' + method,
    method: methods[methodIndex],
    url: url,
    headers: 'x-test: true',
    created: LAST_TIME,
    updated: LAST_TIME
  };
  return obj;
};

DataGenerator.generateRequests = function(size) {
  size = size || 25;
  var result = [];
  for (var i = 0; i < size; i++) {
    result.push(DataGenerator.genRequestObject());
  }
  return result;
};

DataGenerator.generateHistory = function(size) {
  var requests = DataGenerator.generateRequests(size);
  var db = new PouchDB('history-requests');
  return db.bulkDocs(requests);
};

DataGenerator.destroyData = function() {
  var db = new PouchDB('history-requests');
  return db.destroy();
};


demo/index.js
import { html } from 'lit-html';
import { ArcDemoPage } from '@advanced-rest-client/arc-demo-helper/ArcDemoPage.js';
import '@anypoint-web-components/anypoint-checkbox/anypoint-checkbox.js';
import '@anypoint-web-components/anypoint-radio-button/anypoint-radio-button.js';
import '@anypoint-web-components/anypoint-radio-button/anypoint-radio-group.js';
import '@anypoint-web-components/anypoint-button/anypoint-button.js';
import '@advanced-rest-client/arc-demo-helper/arc-interactive-demo.js';
import '@polymer/paper-toast/paper-toast.js';
import { DataGenerator } from '@advanced-rest-client/arc-data-generator/arc-data-generator.js';
import '../history-menu.js';

class DemoPage extends ArcDemoPage {
  constructor() {
    super();
    this.initObservableProperties([
      'draggableEnabled',
      'compatibility',
      'listType',
      'dropValue'
    ]);
    this._componentName = 'history-menu';
    this.demoStates = ['Material', 'Anypoint'];
    this.listType = 'default';

    this._demoStateHandler = this._demoStateHandler.bind(this);
    this._toggleMainOption = this._toggleMainOption.bind(this);
    this._listTypeHandler = this._listTypeHandler.bind(this);
    this.generateData = this.generateData.bind(this);
    this.deleteData = this.deleteData.bind(this);
    this.refreshList = this.refreshList.bind(this);
    this.firstchanged = this.firstchanged.bind(this);
    this.firstDeleted = this.firstDeleted.bind(this);
    this.addNewItem = this.addNewItem.bind(this);
    this._dragoverHandler = this._dragoverHandler.bind(this);
    this._dragleaveHandler = this._dragleaveHandler.bind(this);
    this._dropHandler = this._dropHandler.bind(this);
  }

  _toggleMainOption(e) {
    const { name, checked } = e.target;
    this[name] = checked;
  }

  _demoStateHandler(e) {
    const state = e.detail.value;
    switch (state) {
      case 0:
        this.compatibility = false;
        break;
      case 1:
        this.compatibility = true;
        break;
    }
  }

  _listTypeHandler(e) {
    const { name, checked } = e.target;
    if (!checked) {
      return;
    }
    this.listType = name;
  }

  async generateData() {
    await DataGenerator.insertHistoryRequestData({
      requestsSize: 100
    });
    document.getElementById('genToast').opened = true;
    const e = new CustomEvent('data-imported', {
      bubbles: true
    });
    document.body.dispatchEvent(e);
  }

  async deleteData() {
    await DataGenerator.destroyHistoryData();
    document.getElementById('delToast').opened = true;
    const e = new CustomEvent('datastore-destroyed', {
      detail: {
        datastore: 'all'
      },
      bubbles: true
    });
    document.body.dispatchEvent(e);
  }

  refreshList() {
    document.querySelector('history-menu').refresh();
  }

  firstchanged() {
    let item = document.querySelector('history-menu').requests[0];
    item = Object.assign({}, item);
    item.updated = Date.now();
    const e = new CustomEvent('request-object-changed', {
      detail: {
        request: item,
        type: 'history'
      },
      bubbles: true
    });
    document.body.dispatchEvent(e);
  }

  firstDeleted() {
    const item = document.querySelector('history-menu').requests[0];
    const e = new CustomEvent('request-object-deleted', {
      detail: {
        oldRev: item._rev,
        id: item._id,
        type: 'history'
      },
      bubbles: true
    });
    document.body.dispatchEvent(e);
  }

  addNewItem() {
    const item = DataGenerator.generateHistoryObject();
    item.updated = Date.now();
    const e = new CustomEvent('request-object-changed', {
      detail: {
        request: item,
        type: 'history'
      },
      bubbles: true
    });
    document.body.dispatchEvent(e);
  }

  _dragoverHandler(e) {
    e.preventDefault();
    e.dataTransfer.dropEffect = 'copy';
    e.currentTarget.classList.add('drag-over');
  }

  _dragleaveHandler(e) {
    e.currentTarget.classList.remove('drag-over');
  }

  _dragEnterHandler(e) {
    e.currentTarget.classList.add('drag-over');
  }

  _dropHandler(e) {
    e.preventDefault();
    const data = e.dataTransfer.getData('arc/request-object');
    // format data
    const request = JSON.parse(data);
    this.dropValue = JSON.stringify(request, null, 2);
    console.log(request);
    e.currentTarget.classList.remove('drag-over');
  }

  _demoTemplate() {
    const {
      demoStates,
      darkThemeActive,
      draggableEnabled,
      compatibility,
      listType,
      dropValue
    } = this;
    return html`
      <section class="documentation-section">
        <h3>Interactive demo</h3>
        <p>
          This demo lets you preview the history menu element with various
          configuration options.
        </p>

        <arc-interactive-demo
          .states="${demoStates}"
          @state-chanegd="${this._demoStateHandler}"
          ?dark="${darkThemeActive}"
        >
          <history-menu
            ?draggableEnabled="${draggableEnabled}"
            ?compatibility="${compatibility}"
            .listType="${listType}"
            slot="content"></history-menu>


          <label slot="options" id="mainOptionsLabel">Options</label>
          <anypoint-checkbox
            aria-describedby="mainOptionsLabel"
            slot="options"
            name="draggableEnabled"
            @change="${this._toggleMainOption}"
            >Draggable</anypoint-checkbox
          >

          <label slot="options" id="listTypeLabel">List type</label>
          <anypoint-radio-group
            slot="options"
            selectable="anypoint-radio-button"
            aria-labelledby="listTypeLabel"
          >
            <anypoint-radio-button
              @change="${this._listTypeHandler}"
              checked
              name="default"
              >Default</anypoint-radio-button
            >
            <anypoint-radio-button
              @change="${this._listTypeHandler}"
              name="comfortable"
              >Comfortable</anypoint-radio-button
            >
            <anypoint-radio-button
              @change="${this._listTypeHandler}"
              name="compact"
              >Compact</anypoint-radio-button
            >
          </anypoint-radio-group>
        </arc-interactive-demo>

        ${draggableEnabled ? html`<section
          class="drop-target"
          @dragover="${this._dragoverHandler}"
          @dragleave="${this._dragleaveHandler}"
          @dragenter="${this._dragEnterHandler}"
          @drop="${this._dropHandler}">
          Drop request here
          ${dropValue ? html`<output>${dropValue}</output>` : ''}
        </section>` : ''}

        <div class="data-options">
          <h3>Data options</h3>

          <anypoint-button @click="${this.generateData}">Generate 100 requests</anypoint-button>
          <anypoint-button @click="${this.deleteData}">Clear list</anypoint-button>
          <anypoint-button @click="${this.refreshList}">Refresh list</anypoint-button>
          <anypoint-button @click="${this.firstchanged}">Inform first item changed</anypoint-button>
          <anypoint-button @click="${this.firstDeleted}">Inform first item deleted</anypoint-button>
          <anypoint-button @click="${this.addNewItem}">Add new history item</anypoint-button>
        </div>
      </section>

      <paper-toast id="genToast" text="The request data has been generated"></paper-toast>
      <paper-toast id="delToast" text="The request data has been removed"></paper-toast>
      <paper-toast id="navToast" text="Navigation ocurred"></paper-toast>
    `;
  }

  _introductionTemplate() {
    return html`
      <section class="documentation-section">
        <h3>Introduction</h3>
        <p>
          Advanced REST Client history mennu is a part of application menu.
          It is styled for material design lists with compatibility with
          Anypoint platform.
        </p>
      </section>
    `;
  }

  _usageTemplate() {
    return html`
      <section class="documentation-section">
        <h2>Usage</h2>
        <p>History menu comes with 2 predefied styles:</p>
        <ul>
          <li><b>Material</b> - Normal state</li>
          <li>
            <b>Compatibility</b> - To provide compatibility with Anypoint design
          </li>
        </ul>
      </section>
    `;
  }

  contentTemplate() {
    return html`
      <h2>ARC history menu</h2>
      ${this._demoTemplate()}
      ${this._introductionTemplate()}
      ${this._usageTemplate()}

      <paper-toast id="brightnessAction" text="Turing lights on"></paper-toast>
      <paper-toast id="alarmAction" text="Setting the alarm"></paper-toast>
      <paper-toast id="clearAction" text="Clearing all actions"></paper-toast>
    `;
  }
}

window.addEventListener('navigate', function() {
  document.getElementById('navToast').opened = true;
});

const instance = new DemoPage();
instance.render();
window._demo = instance;


history-menu.d.ts
/**
 * DO NOT EDIT
 *
 * This file was automatically generated by
 *   https://github.com/Polymer/tools/tree/master/packages/gen-typescript-declarations
 *
 * To modify these typings, edit the source file(s):
 *   history-menu.js
 */


// tslint:disable:variable-name Describing an API that's defined elsewhere.

declare namespace UiElements {

  /**
   * A list of history requests in the ARC main menu.
   *
   * The element uses direct implementation of the PouchDB to make a query to the
   * datastore. It also listens to events fired by the `arc-model` elements to
   * update state of the history requests.
   *
   * ### Example
   * ```
   * <history-menu></history-menu>
   * ```
   *
   * ### Sizing the element
   *
   * The element uses `<iron-list>` to render the data in the view. The list is set
   * to be flex vertically. It means that the element has to be sized directly by the
   * hosting application or otherwise it size will be 0px.
   *
   * It can be done using flex layout and making the element to be `flex: 1`.
   *
   * ## Required dependencies
   *
   * The element works with `@advanced-rest-client/arc-models/request-model.js`
   * component. It should be included into the DOM.
   *
   * ### Styling
   * `<history-menu>` provides the following custom properties and mixins for styling:
   *
   * Custom property | Description | Default
   * ----------------|-------------|----------
   * `--history-menu-background-color` | Background color of the menu | `inherit`
   * `--history-menu-selected-post-method-color` | Font color of selected item POST method label | `#fff`
   * `--history-menu-focused-post-method-color` | Font color of focused item POST method label | `rgb(33, 150, 243)`
   * `--history-menu-selected-method-label-background-color` | Bg color of the POST method when focused | `#fff`
   * `--history-menu-selected-item-background-color` | Background color of the selected list item | `#FF9800`
   * `--history-menu-selected-item-color` | Color of the selected list item | `#fff`
   * `--history-menu-history-group-header-font-weigth` | Group header border color | `bold`
   * `--history-menu-history-group-header-border-color` | Group header border color | `#ddd`
   * `--history-menu-history-group-header-color` | Font color of the group header` | `rgba(0, 0, 0, 0.54)`
   * `--arc-menu-empty-info-color` | Color applied to the empty info section | ``
   * `--arc-menu-empty-info-title-color` | Color applied to the title in the empty info section | ``
   */
  class HistoryMenu extends
    RequestsListMixin(
    HistoryListMixin(
    Object)) {

    /**
     * Database ID of the selected item.
     */
    selectedItem: string|null|undefined;

    /**
     * Adds draggable property to the request list item element.
     * The `dataTransfer` object has `arc/request-object` mime type with
     * serialized JSON with request model.
     */
    draggableEnabled: boolean|null|undefined;
    connectedCallback(): void;
    disconnectedCallback(): void;

    /**
     * Computes class name for the HTML element representing a history item.
     */
    _computeItemClass(id: String|null, selectedItem: String|null): String|null;

    /**
     * Called every time the element changed it's scroll position. It will call the `makeQuery`
     * function when there's less than 120px left to scroll. (also it must be opened and must not
     * already querying).
     */
    _scrollHandler(): void;

    /**
     * Notifies the list that the resize event occurred.
     * Should be called whhen content of the list changed but the list wasn't
     * visible at the time.
     */
    notifyResize(): void;

    /**
     * Handler for the `click` event on the item.
     */
    _openHistory(e: ClickEvent|null): void;

    /**
     * Handler for the `dragstart` event added to the list item when `draggableEnabled`
     * is set to true.
     * This function sets request data on the `dataTransfer` object with `arc/request-object`
     * mime type. The request data is a serialized JSON with request model.
     */
    _dragStart(e: Event|null): void;

    /**
     * Computes value for the `draggable` property of the list item.
     * When `draggableEnabled` is set it returns true which is one of the
     * conditions to enable drag and drop on an element.
     *
     * @param draggableEnabled Current value of `draggableEnabled`
     * @returns `true` or `false` (as string) depending on the argument.
     */
    _computeDraggableValue(draggableEnabled: Boolean|null): String|null;
  }
}

declare global {

  interface HTMLElementTagNameMap {
    "history-menu": UiElements.HistoryMenu;
  }
}

export {};


karma.sl.config.js
/* eslint-disable import/no-extraneous-dependencies */
const merge = require('deepmerge');
const slSettings = require('@advanced-rest-client/testing-karma-sl/sl-settings.js');
const createBaseConfig = require('./karma.conf.js');

module.exports = (config) => {
  const cnf = slSettings();
  cnf.sauceLabs.testName = 'history-menu';
  cnf.browsers = [
    'SL_Chrome',
    'SL_Chrome-1',
    'SL_Firefox',
    'SL_Firefox-1',
    // 'SL_Safari',
    // 'SL_EDGE'
  ];
  if (process.env.TRAVIS) {
    const buildLabel = 'TRAVIS #' + process.env.TRAVIS_BUILD_NUMBER + ' (' + process.env.TRAVIS_BUILD_ID + ')';

    cnf.browserStack = {
      build: buildLabel,
      tunnelIdentifier: process.env.TRAVIS_JOB_NUMBER
    };

    cnf.sauceLabs.build = buildLabel;
    cnf.sauceLabs.tunnelIdentifier = process.env.TRAVIS_JOB_NUMBER;
  }

  config.set(merge(createBaseConfig(config), cnf));

  return config;
};


history-menu.js
/**
@license
Copyright 2018 The Advanced REST client authors <arc@mulesoft.com>
Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
*/
import { LitElement, html, css } from 'lit-element';
import { HistoryListMixin } from '@advanced-rest-client/history-list-mixin/history-list-mixin.js';
import { RequestsListMixin } from '@advanced-rest-client/requests-list-mixin/requests-list-mixin.js';
import styles from '@advanced-rest-client/requests-list-mixin/requests-list-styles.js';
import '@anypoint-web-components/anypoint-item/anypoint-icon-item.js';
import '@anypoint-web-components/anypoint-item/anypoint-item-body.js';
import '@polymer/paper-progress/paper-progress.js';
import '@api-components/http-method-label/http-method-label.js';
import { AnypointMenuMixin } from '@anypoint-web-components/anypoint-menu-mixin/anypoint-menu-mixin.js';
// import labelStyles from '@api-components/http-method-label/http-method-label-common-styles.js';

class HistoryMenuWrapper extends AnypointMenuMixin(LitElement) {
  render() {
    return html`<slot></slot>`;
  }
}
window.customElements.define('history-menu-wrapper', HistoryMenuWrapper);
/**
 * A list of history requests in the ARC main menu.
 *
 * The element uses direct implementation of the PouchDB to make a query to the
 * datastore. It also listens to events fired by the `arc-model` elements to
 * update state of the history requests.
 *
 * ### Example
 * ```
 * <history-menu></history-menu>
 * ```
 *
 * ### Sizing the element
 *
 * The element uses `<iron-list>` to render the data in the view. The list is set
 * to be flex vertically. It means that the element has to be sized directly by the
 * hosting application or otherwise it size will be 0px.
 *
 * It can be done using flex layout and making the element to be `flex: 1`.
 *
 * ## Required dependencies
 *
 * The element works with `@advanced-rest-client/arc-models/request-model.js`
 * component. It should be included into the DOM.
 *
 * ### Styling
 * `<history-menu>` provides the following custom properties and mixins for styling:
 *
 * Custom property | Description | Default
 * ----------------|-------------|----------
 * `--history-menu-background-color` | Background color of the menu | `inherit`
 * `--history-menu-selected-post-method-color` | Font color of selected item POST method label | `#fff`
 * `--history-menu-focused-post-method-color` | Font color of focused item POST method label | `rgb(33, 150, 243)`
 * `--history-menu-selected-method-label-background-color` | Bg color of the POST method when focused | `#fff`
 * `--history-menu-selected-item-background-color` | Background color of the selected list item | `#FF9800`
 * `--history-menu-selected-item-color` | Color of the selected list item | `#fff`
 * `--history-menu-history-group-header-font-weigth` | Group header border color | `bold`
 * `--history-menu-history-group-header-border-color` | Group header border color | `#ddd`
 * `--history-menu-history-group-header-color` | Font color of the group header` | `rgba(0, 0, 0, 0.54)`
 * `--arc-menu-empty-info-color` | Color applied to the empty info section | ``
 * `--arc-menu-empty-info-title-color` | Color applied to the title in the empty info section | ``
 *
 * @customElement
 * @memberof UiElements
 * @demo demo/index.html
 * @appliesMixin RequestsListMixin
 * @appliesMixin HistoryListMixin
 */
class HistoryMenu extends HistoryListMixin(RequestsListMixin(LitElement)) {
  static get styles() {
    return [
      styles,
      css`:host {
        display: block;
        background-color: var(--history-menu-background-color, inherit);
        position: relative;
        display: flex;
        flex-direction: column;
      }

      .list {
        overflow: auto;
      }

      .url {
        white-space: nowrap;
        text-overflow: ellipsis;
        overflow: hidden;
        font-size: 14px;
      }

      paper-progress {
        width: calc(100% - 32px);
        margin: 0 16px;
        position: absolute;
      }

      .empty-info {
        font-size: var(--arc-font-body1-font-size);
        font-weight: var(--arc-font-body1-font-weight);
        line-height: var(--arc-font-body1-line-height);
        font-style: italic;
        margin: 1em 16px;
        color: var(--arc-menu-empty-info-color);
      }

      .empty-title {
        white-space: var(--arc-font-nowrap-white-space);
        overflow: var(--arc-font-nowrap-overflow);
        text-overflow: var(--arc-font-nowrap-text-overflow);
        font-size: var(--arc-font-title-font-size);
        font-weight: var(--arc-font-title-font-weight);
        line-height: var(--arc-font-title-line-height);
        white-space: normal;
        color: var(--arc-menu-empty-info-title-color);
      }

      .empty-message {
        flex: 1;
        flex-basis: 0.000000001px;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;
        text-align: center;
      }

      .empty-state-image {
        width: 180px;
        height: 120px;
      }

      [hidden] {
        display: none !important;
      }

      .selected {
        color: var(--primary-color);
      }`
    ];
  }

  _unavailableTemplate() {
    return html`<div class="empty-message">
      <h3 class="empty-title">Send a request and recall it from here</h3>
      <p class="empty-info">Once you made a request it will appear in this place.</p>
    </div>`;
  }

  _listTemplate() {
    const items = this.requests || [];
    const { draggableEnabled, _hasTwoLines, compatibility } = this;
    return items.map((item, index) => html`
      ${item.hasHeader ? html`<div class="history-group-header">${item.header}</div>` : ''}
      <anypoint-icon-item
        data-index="${index}"
        data-id="${item._id}"
        @click="${this._openHistory}"
        class="request-list-item"
        draggable="${draggableEnabled ? 'true' : 'false'}"
        @dragstart="${this._dragStart}"
        tabindex="-1"
        title="${item.url}"
        role="menuitem"
        ?compatibility="${compatibility}">
        <http-method-label
          method="${item.method}"
          slot="item-icon"></http-method-label>
        <anypoint-item-body
          ?twoline="${_hasTwoLines}"
          ?compatibility="${compatibility}">
          <div class="url">${item.url}</div>
          <div secondary="">${item.timeLabel}</div>
        </anypoint-item-body>
      </anypoint-icon-item>`);
  }

  render() {
    const { dataUnavailable, hasRequests, querying, selectedItem } = this;
    return html`
    ${this.modelTemplate}
    <paper-progress ?hidden="${!querying}" indeterminate></paper-progress>
    ${dataUnavailable ? this._unavailableTemplate() : ''}
    <history-menu-wrapper
      class="list"
      selectable="anypoint-icon-item"
      attrforselected="data-id"
      .selected="${selectedItem}"
      @selected-changed="${this._selectionChanged}">
      ${hasRequests ? this._listTemplate() : ''}
    </history-menu-wrapper>`;
  }

  static get properties() {
    return {
      // Database ID of the selected item.
      selectedItem: { type: String },
      /**
       * Adds draggable property to the request list item element.
       * The `dataTransfer` object has `arc/request-object` mime type with
       * serialized JSON with request model.
       */
      draggableEnabled: { type: Boolean },
      /**
       * Enables compatibility with Anypoint platform
       */
      compatibility: { type: Boolean }
    };
  }

  get _list() {
    if (!this.__list) {
      this.__list = this.shadowRoot.querySelector('.list');
    }
    return this.__list;
  }

  constructor() {
    super();
    this._scrollHandler = this._scrollHandler.bind(this);
  }

  connectedCallback() {
    if (super.connectedCallback) {
      super.connectedCallback();
    }
    this.type = 'history';
    this._addScrollEvent();
  }

  disconnectedCallback() {
    if (super.disconnectedCallback) {
      super.disconnectedCallback();
    }
    this._list.removeEventListener('scroll', this._scrollHandler);
  }

  firstUpdated() {
    this._addScrollEvent();
  }

  _addScrollEvent() {
    const list = this._list;
    if (!list) {
      return;
    }
    list.addEventListener('scroll', this._scrollHandler);
  }
  /**
   * Called every time the element changed it's scroll position. It will call the `makeQuery`
   * function when there's less than 120px left to scroll. (also it must be opened and must not
   * already querying).
   */
  _scrollHandler() {
    if (this.querying) {
      return;
    }
    const elm = this._list;
    const delta = elm.scrollHeight - (elm.scrollTop + elm.offsetHeight);
    if (delta < 120) {
      this.loadNext();
    }
  }
  /**
   * Handler for the `click` event on the item.
   * @param {ClickEvent} e
   */
  _openHistory(e) {
    const index = Number(e.currentTarget.dataset.index);
    const request = this.requests[index];
    const id = request._id;
    this._openRequest(id);
  }
  /**
   * Handler for the `dragstart` event added to the list item when `draggableEnabled`
   * is set to true.
   * This function sets request data on the `dataTransfer` object with `arc/request-object`
   * mime type. The request data is a serialized JSON with request model.
   * @param {Event} e
   */
  _dragStart(e) {
    if (!this.draggableEnabled) {
      return;
    }
    const index = Number(e.currentTarget.dataset.index);
    const request = this.requests[index];
    const data = JSON.stringify(request);
    e.dataTransfer.setData('arc/request-object', data);
    e.dataTransfer.setData('arc/history-request', request._id);
    e.dataTransfer.setData('arc-source/history-menu', request._id);
    e.dataTransfer.effectAllowed = 'copy';
  }

  _selectionChanged(e) {
    this.selectedItem = e.detail.value;
  }
}
window.customElements.define('history-menu', HistoryMenu);
